Source/RunActivity/Viewer3D/Debugging/TimetableWindow.cs

```diff
@@ -0,0 +1,928 @@
+			F.trainBrush = new SolidBrush(Color.Red);
+			F.InactiveTrainBrush = new SolidBrush(Color.DarkRed);
+			F.sidingBrush = new SolidBrush(Color.Blue);
+			F.PlatformBrush = new SolidBrush(Color.DarkBlue);
+			F.SignalBrush = new SolidBrush(Color.DarkRed);
+			F.pbCanvas.BackColor = Color.FromArgb(250, 240, 230);
+		}
+
+		private void AdjustControlLocations()
+		{
+			if (F.Height < 600 || F.Width < 800) return;
+
+			if (F.oldHeight != F.Height || F.oldWidth != F.Width) //use the label "Res" as anchor point to determine the picture size
+			{
+				F.oldWidth = F.Width; F.oldHeight = F.Height;
+
+				F.pbCanvas.Top = 50;
+				F.pbCanvas.Width = F.label1.Left - 25;                  // 25 pixels found by trial and error
+				F.pbCanvas.Height = F.Height - F.pbCanvas.Top - 45;  // 45 pixels found by trial and error
+
+				if (F.pbCanvas.Image != null)
+					F.pbCanvas.Image.Dispose();
+				F.pbCanvas.Image = new Bitmap(F.pbCanvas.Width, F.pbCanvas.Height);
+			}
+			if (F.firstShow)
+			{
+				// Center the view on the player's locomotive
+				var pos = Program.Simulator.PlayerLocomotive.WorldPosition;
+				var ploc = new PointF(pos.TileX * 2048 + pos.Location.X, pos.TileZ * 2048 + pos.Location.Z);
+#pragma warning disable CS1690 // Accessing a member on a field of a marshal-by-reference class may cause a runtime exception
+                F.ViewWindow.X = ploc.X - F.minX - F.ViewWindow.Width / 2;
+				F.ViewWindow.Y = ploc.Y - F.minY - F.ViewWindow.Width / 2;
+#pragma warning restore CS1690 // Accessing a member on a field of a marshal-by-reference class may cause a runtime exception
+				F.firstShow = false;
+			}
+
+			// Sufficient to accommodate the whole route plus 50%
+			var xRange = F.maxX - F.minX;
+			var yRange = F.maxY - F.minY;
+			var maxSize = (int)(((xRange > yRange) ? xRange : yRange) * 1.5);
+			F.windowSizeUpDown.Maximum = (decimal)maxSize;
+		}
+
+		public void PopulateItemLists()
+		{
+			foreach (var item in F.simulator.TDB.TrackDB.TrItemTable)
+            {
+				switch (item.ItemType)
+				{
+					case TrItem.trItemType.trSIGNAL:
+						if (item is SignalItem)
+						{
+							SignalItem si = item as SignalItem;
+
+							if (si.SigObj >= 0 && si.SigObj < F.simulator.Signals.SignalObjects.Length)
+							{
+								var s = F.simulator.Signals.SignalObjects[si.SigObj];
+								if (s != null && s.isSignal && s.isSignalNormal())
+									F.signals.Add(new SignalWidget(si, s));
+							}
+						}
+						break;
+
+					case TrItem.trItemType.trSIDING:
+						// Sidings have 2 ends but are not always listed in pairs in the *.tdb file
+						// Neither are their names unique (e.g. Bernina Bahn).
+						// Find whether this siding is a new one or the other end of an old one.
+						// If other end, then find the right-hand one as the location for a single label.
+						var oldSidingIndex = F.sidings.FindIndex(r => r.LinkId == item.TrItemId && r.Name == item.ItemName);
```

Comment:
This is a double for-loop, and therefore, an O(n^2) operation. This may cause slowdowns with large numbers of visible sidings. Consider changing the `sidings` variable of the debug viewer to a dictionary, like so:
```C#
public IDictionary<(uint, string), SidingWidget> RenderedSidings { get; } = new Dictionary<(uint, string), SidingWidget>();
```
Dictionaries can perform these lookups in constant time and would make this operation O(n).

Reply:
True and well spotted. However, it happens only once at start-up not every cycle so I prefer to keep it simple.